pkg/api/api.go

```diff
@@ -118,47 +113,59 @@ func (a API) ReceiveEvent(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	// TODO: Implement key matching from core data loader.
-
-	body, err := io.ReadAll(io.LimitReader(r.Body, int64(a.config.EventAPI.MaxSize)))
-	if err != nil {
-		a.writeResponse(w, apiResponse{
-			StatusCode: http.StatusBadRequest,
-			Error:      "Could not read event payload",
-		})
-		return
-	}
+	// Create a new channel which receives a stream of events from the incoming HTTP request
+	byteStream := make(chan json.RawMessage)
+	eg := errgroup.Group{}
+	eg.Go(func() error {
+		return eventstream.ParseStream(ctx, r.Body, byteStream, a.config.EventAPI.MaxSize)
+	})
 
-	events, err := parseBody(body)
-	if err != nil {
-		a.writeResponse(w, apiResponse{
-			StatusCode: http.StatusBadRequest,
-			Error:      "Unable to process event payload",
-		})
-		return
-	}
+	// Create a new channel which holds all event IDs as a slice.
+	var (
+		ids    = []string{}
+		idChan = make(chan string)
+	)
+	eg.Go(func() error {
+		for item := range idChan {
+			ids = append(ids, item)
+		}
+		return nil
+	})
 
-	eg := &errgroup.Group{}
-	for _, evt := range events {
-		copied := evt
-		eg.Go(func() error {
-			if err := a.handler(r.Context(), copied); err != nil {
-				a.log.Error().Str("event", copied.Name).Err(err).Msg("error handling event")
+	// Process those incoming events
+	eg.Go(func() error {
+		// TODO: Iterate through event stream and process event.
+		for byt := range byteStream {
+			evt := event.Event{}
+			if err := json.Unmarshal(byt, &evt); err != nil {
 				return err
 			}
-			return nil
-		})
-	}
+			id, err := a.handler(r.Context(), &evt)
+			if err != nil {
+				a.log.Error().Str("event", evt.Name).Err(err).Msg("error handling event")
+				return err
+			}
+			idChan <- id
+		}
+
+		// Close the idChan so that we stop appending to the ID slice.
+		close(idChan)
+		return nil
+	})
 
 	if err := eg.Wait(); err != nil {
-		a.writeResponse(w, apiResponse{
-			StatusCode: http.StatusBadRequest,
-			Error:      err.Error(),
+		w.WriteHeader(400)
+		_ = json.NewEncoder(w).Encode(apiutil.EventAPIResponse{
+			IDs:    ids,
```

Comment:
the `ids` here mean that we're returning the IDs that have been successfully processed?

Reply:
Correct!